Sources/SourceKitLSP/Swift/SyntaxHighlightingToken.swift

```diff
@@ -0,0 +1,453 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SourceKitD
+import LanguageServerProtocol
+import LSPLogging
+
+/// A ranged token in the document used for syntax highlighting.
+public struct SyntaxHighlightingToken: Hashable {
+  /// The range of the token in the document. Must be on a single line.
+  public var range: Range<Position> {
+    didSet {
+      assert(range.lowerBound.line == range.upperBound.line)
+    }
+  }
+  /// The token type.
+  public var kind: Kind
+  /// Additional metadata about the token.
+  public var modifiers: Modifiers
+
+  /// The (inclusive) start position of the token.
+  /// Setting it shifts the token and preserves the length.
+  public var start: Position {
```

Comment:
I agree, this behavior is a bit subtle. The main advantage of using a property setter here is that it makes shifting the tokens e.g. by line very convenient, i.e. we can write

```swift
token.start.line += ...
```

without having to construct new positions manually. Previously tokens were defined in terms of start + length and this setter basically preserves those semantics.

But I can see how it would be better to make this more explicit with a method.

Reply:
Just an idea: `move(lineDelta: Int, utf16IndexDelta: Int? = nil)`